test/router-luatest/reload_test.lua

```diff
@@ -0,0 +1,265 @@
+        ivshard.router.callrw(bucket_id, 'insert', {'test', {1, bucket_id}})
+        ivshard.router.callrw(bucket_id, 'vshard.storage.sync', {})
+        local res, err = ivshard.router.callro(bucket_id, 'get', {'test', {1}})
+        ilt.assert_equals(err, nil)
+        ilt.assert_equals(res, {1, bucket_id})
+    end)
+end
+
+g.test_basic = function(g)
+    -- Latest meaningful commit:
+    --     "router: fix reload problem with global function refs".
+    local hash = '139223269cddefe2ba4b8e9f6e44712f099f4b35'
+    local router = create_router_at(hash)
+    test_basic_template(router)
+    reload_router(router)
+    test_basic_template(router)
+    vtest.drop_instance(g, router)
+end
+
+local function test_discovery_template(g, router)
+    -- We need to wait, until all of the services are indeed reloaded or started
+    t.helpers.retrying({timeout = vtest.wait_timeout}, function()
+        if not router:grep_log('discovery_f has been started', 1024) then
+            error("Discovery haven't started yet")
+        end
+    end)
+
+    -- Working with second replicaset (1 instance)
+    local uuid = g.replica_2_a:exec(function()
+        return box.info.cluster.uuid
+    end)
+
+    router:exec(function(uuid)
+        -- Everything is all right
+        local router = ivshard.router.internal.static_router
+        local discovery = router.discovery_service
+        ivtest.service_wait_for_ok(discovery,
+            {on_yield = ivshard.router.discovery_wakeup})
+        ilt.assert_equals(router.known_bucket_count, 100)
+
+        -- Break connection and the number of known buckets
+        local conn = router.replicasets[uuid].master.conn
+        local known = router.known_bucket_count
+        router.replicasets[uuid].master.conn = nil
+        router.known_bucket_count = known - 1
+
+        ivtest.service_wait_for_error(discovery, 'Error during discovery',
+            {on_yield = ivshard.router.discovery_wakeup})
+
+        -- Restore everything
+        router.replicasets[uuid].master.conn = conn
+        router.known_bucket_count = known
+
+        -- It's all right again
+        ivtest.service_wait_for_ok(discovery,
+            {on_yield = ivshard.router.discovery_wakeup})
+    end, {uuid})
+end
+
+g.test_discovery = function(g)
+    -- Service_info was introduced in the following commit:
+    --     "router: add saving of background service statuses".
+    local hash = 'f5f386a5a35e6e5efd8f4f2ed1b3d208fdae9095'
+    local router = create_router_at(hash)
+    test_discovery_template(g, router)
+    reload_router(router)
+    test_discovery_template(g, router)
+    vtest.drop_instance(g, router)
+end
+
+local function test_master_search_template(g, router, auto_master_cfg)
+    vtest.router_cfg(router, auto_master_cfg)
+
+    -- Working with first replicaset (2 instances)
+    local rs_uuid, master_uuid = unpack(g.replica_1_a:exec(function()
+        local info = box.info()
+        return {info.cluster.uuid, info.uuid}
+    end))
```

Comment:
Wouldn't it work like this (without unpack)?
```Lua
    local rs_uuid, master_uuid = g.replica_1_a:exec(function()
        local info = box.info()
        return info.cluster.uuid, info.uuid
    end)
```

Reply:
It would